components/org.wso2.carbon.identity.application.authenticator.google/src/main/java/org/wso2/carbon/identity/application/authenticator/google/GoogleOAuth2Authenticator.java

```diff
@@ -51,13 +71,183 @@ public class GoogleOAuth2Authenticator extends OpenIDConnectAuthenticator {
+     * @return Integrity of the authentication request sent via Google One Tap.
+     */
+    private boolean validateCSRFCookies(HttpServletRequest request) {
+
+        if (request.getCookies() == null) {
+            if (log.isDebugEnabled()) {
+                log.debug("No valid cookie found for Google One Tap authentication.");
+            }
+            return false;
+        }
+        List<Cookie> crossRefCookies = Arrays.stream(request.getCookies())
+                .filter(cookie -> cookie.getName().equalsIgnoreCase(G_CSRF_TOKEN))
+                .collect(Collectors.toList());
+
+        if (crossRefCookies.isEmpty() || crossRefCookies.get(0) == null) {
+            if (log.isDebugEnabled()) {
+                log.debug("No CSRF cookie found. Invalid request.");
+            }
+            return false;
+        }
+        String crossRefCookieHalf = crossRefCookies.get(0).getValue();
+        String crossRefParamHalf = request.getParameter(G_CSRF_TOKEN);
+
+        if (StringUtils.isEmpty(crossRefParamHalf) || StringUtils.isEmpty(crossRefCookieHalf)) {
+            if (log.isDebugEnabled()) {
+                log.debug("No CSRF parameter found. Invalid request.");
+            }
+            return false;
+        }
+        if (!crossRefParamHalf.equals(crossRefCookieHalf)) {
+            if (log.isDebugEnabled()) {
+                log.debug("CSRF validation failed for Google One Tap.");
+            }
+            return false;
+        }
+        if (log.isDebugEnabled()) {
+            log.debug("Validated CSRF cookies successfully for Google One Tap.");
+        }
+        return true;
+    }
+
+    @Override
+    protected String mapIdToken(AuthenticationContext context, HttpServletRequest request,
+                                OAuthClientResponse oAuthResponse) throws AuthenticationFailedException{
+
+        /*
+          Validity of the
+          1. CSRF cookies
+          2. JWT token
+          decide the ability of handling the authentication request in Google One Tap flow.
+         */
+        if (isOneTapEnabled(request)) {
+            Map<String, String> authenticatorProperties = context.getAuthenticatorProperties();
+            String clientID = authenticatorProperties.get(OIDCAuthenticatorConstants.CLIENT_ID);
+
+            boolean validCookies = validateCSRFCookies(request);
+            if (!validCookies) {
+                log.error("CSRF cookie validation failed in Google one tap. Authenticator : "+ getName() +
+                        " Client Id : " + clientID );
+                throw new AuthenticationFailedException(OIDCErrorConstants.
+                        ErrorMessages.INTEGRITY_VIOLATION_EXCEPTION.getMessage());
+            }
+
+            boolean validJWT = validateJWTFromGOT(request, clientID);
+            if (!validJWT){
+                log.error("JWT validation failed in Google one tap. Authenticator : "+ getName() +
```

Comment:
Is it okay to log the client Id?

Reply:
As long as it is not highly sensitive, I think it should be okay for logs. For multi authenticator scenario, there should be a way to clearly identify the exact client ID for an error like this.